### ISSUE TITLE
Run formatter under project root directory

### ISSUE THREAD

**Issue by shouya** [raxod502/apheleia#30]
I am able to use this package mostly successfully with my Elixir code with the following config:

```emacs-lisp
  (add-to-list 'apheleia-formatters '(mix . ("mix" "format" "-")))
  (add-to-list 'apheleia-mode-alist '(elixir-mode . mix))
```

There is one issue though. When I format a file under a project that contains a `.formatter.exs` config, the formatter configuration is not respected. In the documentation of `mix format` it states that

> The formatter will read a .formatter.exs in the current directory for formatter configuration.

So it seems like it was caused by that `mix format -` is not executed in the project directory but the directory of the current buffer file.

Is there a way to specify apheleia to run the formatter under the project directory? The project root directory can be found by either `(project-root (project-current))` or `(projectile-project-root)`.

**Comment by shouya** [comment]
Also, `mix format` command supports specifying `.formatter.exs` via cli option `--dot-formatter`.

>  • --dot-formatter - path to the file with formatter configuration.
>    Defaults to .formatter.exs if one is available. See the "Formatting
>    options" section above for more information.

So the same effect should also be achieved by allowing customizable dynamic arguments in `apheleia-formatters`.

**Comment by mattiaslundberg** [comment]
I ended up with something like this (converted from using doom macros and not tested, my used version [here](https://github.com/mattiaslundberg/dotfiles/blob/main/.doom.d/config.el#L40) ):
```elisp
(defun custom-apheleia-format-buffer (orig-fn command &optional callback)
  (if (string-equal "mix" (car command))
    (let ((default-directory (locate-dominating-file ".formatter.exs" default-directory)))
      (funcall orig-fn command callback))
    (funcall orig-fn command callback))))
(add-function #'apheleia-format-buffer :around #'custom-apheleia-format-buffer)
```

From my testing this locates the correct config to use independently of how the project is structured (for example an umbrella app can have multiple formatting configs that are not necessarily the same).

**Comment by shouya** [comment]
@mattiaslundberg Thank you for the fix - it's working great for me. I personally also adapted the code a bit to make it use `project-root` by default.


```emacs-lisp
  (defun shou/fix-apheleia-project-dir (orig-fn &rest args)
    (let ((project (project-current)))
      (if (not (null project))
          (let ((default-directory (project-root project))) (apply orig-fn args))
        (apply orig-fn args))))

  (advice-add 'apheleia-format-buffer :around #'shou/fix-apheleia-project-dir)
```

^ Here's my [adapted version](https://github.com/shouya/emacs.d/commit/dc819e9c05f4053a90cafbb1cceae05c8b152149) just in case it can be helpful for any one.

**Comment by raxod502** [comment]
I've implemented this feature; let me know if it works for you. Thanks for the suggestion!

**Comment by shouya** [comment]
I can confirm it's now working as expected with the current master branch. Thank you!

**Comment by raxod502** [comment]
Some issues with this feature were brought up in https://github.com/raxod502/apheleia/issues/36, so I reverted it for now. I think the most appropriate thing to do would be to add a new symbol which can be added to a formatter command line, such that if it's present then the formatter is run in the project root. Then we will have the best of both worlds.

**Comment by apeschar** [comment]
The problem with this is/was that formatter configuration not in the project root directory got ignored.  Eg, if you have `.prettierrc` in `myproject/src`, and formatter is executed with `myproject` as its working directory, the config won't be found.

What should (IMHO) be done instead is that in case of formatters that do not automatically detect their configuration in parent directories (ie, `mix format`), apheleia looks for the parent directory containing the relevant config file, in this case `.formatter.exs`. I don't think that this should necessarily have to be the project root.

So perhaps we need a hook or a setting that can use [`locate-dominating-file`](https://doc.endlessparentheses.com/Fun/locate-dominating-file.html) to determine the `current-directory` for particular formatters based on `apheleia-formatter`.  This should work:

~~~
(locate-dominating-file buffer-file-name ".formatter.exs")
~~~

**Comment by raxod502** [comment]
OK, that seems reasonable. Ideally, we could make it possible to specify this configuration in the command line for a formatter (i.e., in `apheleia-formatters`). For example:

```elisp
(setf (alist-get mix apheleia-formatters)
      '((default-directory
          (locate-dominating-file buffer-file-name ".formatter.exs"))
        "mix" "format"))
```

**Comment by shouya** [comment]
Or maybe we can extend this idea further by creating an alist for each formatter to either use file directory or project directory. Well maybe it would be good if we can make everything configurable, something like this:

```emacs-lisp
(setq formatters
   '((mix :workdir project
          :command ("mix" "format" "-")
          :mode elixir-mode)
     (black ....
```

If this can become the case, I really hope that one can specify a user function for the `:command`/`:workdir` to customize according to their need.

**Comment by apeschar** [comment]
`apheleia-formatters` is now a alist mapping formatter names to commands:

~~~
((black "black" "-")
 (brittany "brittany" file)
 (prettier npx "prettier" "--stdin-filepath" filepath)
 (mix "mix" "format" "-"))
~~~

Mode is already mapped by `apheleia-mode-alist`. So I guess the simplest way to extend the current configuration is as follows:

~~~
((black "black" "-")
 (brittany "brittany" file)
 (prettier npx "prettier" "--stdin-filepath" filepath)
 (mix "mix" "format" "-" :workdir (dominating-file ".formatter.exs")))
~~~

Where `:workdir` can be

* `file` (for the default), 
* `project` (for the project root),
* or `(dominating-file "whatever")` for the dominating file.

I prefer this notation to the `(default-directory ...)` notation and it's also consistent with other configuration alists like the one used by straight. Thoughts?

**Comment by raxod502** [comment]
Yeah, that seems like a good idea, and easier to understand than what I proposed. I support it :)

**Comment by joshforged** [comment]
Wouldn't a simpler and more general solution be to just evaluate any non-string args in the command so we can put arbitrary lisp code in it? I have a quick proof of concept PR here if that sounds reasonable to you: https://github.com/raxod502/apheleia/pull/50

**Comment by raxod502** [comment]
That would be okay to support as well, but it doesn't fully address the issue as reported, since it doesn't allow you to change the actual working directory. If your formatter allows you to pass the desired directory on the command line, it's fine, but I expect that this won't be the case for all formatters.

**Comment by merrickluo** [comment]
> OK, that seems reasonable. Ideally, we could make it possible to specify this configuration in the command line for a formatter (i.e., in `apheleia-formatters`). For example:
> 
> ```emacs-lisp
> (setf (alist-get mix apheleia-formatters)
>       '((default-directory
>           (locate-dominating-file buffer-file-name ".formatter.exs"))
>         "mix" "format"))
> ```

Hi, just find out mix format does not work well with apheleia, and this workaround does not work anymore, is there another workaround without define a complicated custom function?

**Comment by raxod502** [comment]
The workaround you quoted was never a working workaround, it was just a proposed syntax for a new feature (which hasn't been implemented yet). Have you tried https://github.com/radian-software/apheleia/issues/30#issuecomment-778150037?

**Comment by merrickluo** [comment]
ah :cry: sorry for not reading correctly, https://github.com/radian-software/apheleia/issues/30#issuecomment-778150037 works, thought there is a solution without modifying the global behavior.
Thanks, I'll use it until the new syntax is implemented.
